api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+
+## Drreg Support For Multi-phase Reservations
+
+Owner: [Abhinav Sharma](https://github.com/abhinav92003)
+
+Upstream issue: [DynamoRIO/dynamorio#3823](https://github.com/DynamoRIO/dynamorio/issues/3823)
+
+Drreg is DynamoRIO’s register reservation framework. It allows users to reserve a reg to use as scratch. Internally, drreg automatically performs the following functions so that the user does not need to. Drreg
+- keeps all required book-keeping like the spill slot to spilled register mapping
+- restores spilled registers to their application value before they are read by an application instruction; also, it re-spills the spilled registers if they are written by an application instruction.
+- performs application state restoration on state restore events like encountering an application fault, and DR detach.
+
+While expanding a scatter or gather instruction in the app2app phase, we need a scratch register to hold the scalar values and masks. In later phases (like the insertion or the instru2instru phase), drcachesim and other DR clients may also use drreg to get scratch registers for their instrumentation.
+
+Drreg initially supported only single phase use and did not keep any state between phases. To support multi-phase use of drreg, we need to solve the following:
+- avoid spill slot conflict across multiple phases: multi-phase use can potentially lead to spill slot conflicts if the same slot is selected in multiple phases. This may clobber the spilled application value and cause the application to crash or otherwise fail.
+- allow aflags spill to any slot: drreg hardcodes the aflags spill slot as the zero-th slot, to simplify some logic. To support the ability to spill aflags in multiple phases, drreg should be able to use any spill slot for aflags.
+- application state restore logic: on a state restore event, we should be able to figure out which slot contains each spilled register's app value. This is complicated by the fact that registers may be spilled by instrumentation added by multiple phases, and the spill regions may overlap which causes the spilled application value to be moved between spill slots.
+
+We explored the following ideas to avoid spill slot conflicts in drreg:
+
+#### Disjoint slot spaces or arenas
```

Comment:
The sections jump from ## to ####.  Best to add a ### in between, parallel to the ### below.

Reply:
I used "####" because I didn't want this showing up as a huge heading.

Reply:
Prior replies were on phone to email so showed up below; for completeness duplicating here:

I would still add a ### up above to keep a strict hierarchy and not have all this content in the "intro" for the ## instead of a full-fledged ### parallel to the one below.  What does the contents show without it, especially the sidebar menu: not sure what the JavaScript does with missing levels.  Please make sure the menus are sane.

Reply:
I removed the markdown heading and just made that text bold. It's better to not have this in the sidebar menu.